draft-ietf-mls-protocol.md

```diff
@@ -2208,33 +2208,38 @@ specifically:
 To allow group members to verify that they agree on the public cryptographic state
 of the group, this section defines a scheme for generating a hash value (called
 the "tree hash") that represents the contents of the group's ratchet tree and the
-members' leaf nodes. The tree hash of a tree is the tree hash of its root node,
-which we define recursively, starting with the leaves.
+members' leaf nodes.
 
-The tree hash of a leaf node is the hash of leaf's `LeafNodeHashInput` object which
-might include a `LeafNode` object depending on whether or not it is blank.
+The tree hash of an individual node is the hash of the node's `TreeHashInput`
+object, which may contain either a `LeafNodeHashInput` or a
+`ParentNodeHashInput` depending on the type of node. `LeafNodeHashInput` objects
+contain the `leaf_index` and the `LeafNode` (if any). `ParentNodeHashInput`
+objects contain the `ParentNode` (if any) and the tree hash of the node's left
+and right children.
 
 ~~~ tls
+struct {
+  NodeType node_type;
```

Comment:
This is defined all the way down in the ratchet tree extension.  Please move it up to this section.